cmd/api/src/api/v2/apiclient/apiclient.go

```diff
@@ -78,6 +78,51 @@ func ReadAPIError(response *http.Response) error {
 	return apiError
 }
 
+func (s Client) ZipRequest(method, path string, params url.Values, body []byte) (*http.Response, error) {
+	endpoint := api.URLJoinPath(s.ServiceURL, path)
+	endpoint.RawQuery = params.Encode()
+
+	if request, err := http.NewRequest(method, endpoint.String(), nil); err != nil {
+		return nil, err
+	} else {
+		// query the Request and hand the response back to the user
+		const (
+			sleepInterval = time.Second * 5
+			maxSleep      = sleepInterval * 5
+		)
+
+		started := time.Now()
+
+		for {
+			// Serialize the Request body - we expect a JSON serializable object here
+			// This must be done on every retry, otherwise the buffer will be empty because it had been read
+			if body != nil {
+				request.Header.Set(headers.ContentType.String(), mediatypes.ApplicationZip.String())
+				request.Body = io.NopCloser(bytes.NewReader(body))
+			}
+
+			// Set our credentials either via signage or bearer token session
+			// Credentials also have to be set with every attempt due to request signing
+			if s.Credentials != nil {
+				if err := s.Credentials.Handle(request); err != nil {
+					return nil, err
+				}
+			}
```

Comment:
Is there any danger of leaving the request body reader open if we fail out of ZipRequest due to bad credentials?

Reply:
I'm not 100% positive, but this is all in integration tests so not the biggest deal